pkg/pillar/zedcloud/authen.go

```diff
@@ -379,73 +380,128 @@ func checkMimeProtoType(r *http.Response) bool {
 }
 
 // VerifySigningCertChain - verify signing certificate chain from controller
-func VerifySigningCertChain(ctx *ZedCloudContext, content []byte) ([]byte, error) {
+func VerifySigningCertChain(ctx *ZedCloudContext, content []byte,
+	checkEcdh bool) error {
 	sm := &zcert.ZControllerCert{}
 	err := proto.Unmarshal(content, sm)
 	if err != nil {
 		log.Errorf("SaveCertsToFile: Unmarshal err %v\n", err)
-		return nil, err
+		return err
 	}
 
-	var block *pem.Block
-	var leafcert *x509.Certificate
-	var certByte []byte
-	var keyCnt int
+	// prepare intermediate certs and validate the payload
+	var keyCnt, signKeyCnt int
 	interm := x509.NewCertPool()
 	for _, cert := range sm.GetCerts() {
 		keyCnt++
-		if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE {
+		switch cert.Type {
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE:
 			ok := interm.AppendCertsFromPEM(cert.GetCert())
 			if !ok {
-				err := fmt.Errorf("VerifySigningCertChain: fail to append intermediate certs")
-				return nil, err
+				errStr := fmt.Sprintf("intermediate cert append fail")
+				log.Errorf("VerifySigningCertChain: " + errStr)
+				return errors.New(errStr)
 			}
-			log.Debugf("VerifySigningCertChain: get intermediate cert %d, len %d\n",
-				keyCnt, len(cert.GetCert()))
-		} else if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING {
-			if len(certByte) > 0 {
-				err := fmt.Errorf("VerifySigningCertChain: received more than one leaf cert")
-				return nil, err
+
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
+			signKeyCnt++
+			if signKeyCnt > 1 {
+				errStr := fmt.Sprintf("received more than one signing cert")
+				log.Errorf("VerifySigningCertChain: " + errStr)
+				return errors.New(errStr)
 			}
-			certByte = cert.GetCert()
-			block, _ = pem.Decode(certByte)
-			if block == nil {
-				err := fmt.Errorf("VerifySigningCertChain: can not get cert block")
-				return nil, err
+
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE:
+			// nothing to do
+
+		default:
+			errStr := fmt.Sprintf("received cert with unknown type %d", cert.Type)
+			log.Errorf("VerifySigningCertChain: " + errStr)
+			return errors.New(errStr)
+		}
+	}
+
+	log.Debugf("VerifySigningCertChain: key count %d\n", keyCnt)
+	if signKeyCnt == 0 {
+		err := fmt.Errorf("VerifySigningCertChain: fail to acquire signing cert")
+		return err
+	}
+
+	// verify signature of certificates
+	for _, cert := range sm.GetCerts() {
+		switch cert.Type {
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
```

Comment:
I think CONTROLLER_SIGNING and ECDH_EXCHANGE should be done in the same case statement; they need identical verification.

Reply:
Sure, Changed